### PR TITLE
Add types to layout/Error

### PR DESCRIPTION
**Pull request by iamstiil**
Required #4963

### CODE REVIEW

File: packages/ra-ui-materialui/src/layout/Error.tsx [hunk 1]
```diff
@@ -1,5 +1,5 @@
-import * as React from 'react';
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Error.tsx:1]: please revert this change.

File: packages/ra-ui-materialui/src/layout/Error.tsx [hunk 2]
```diff
@@ -100,12 +100,37 @@ const Error = props => {
...
+export declare type ErrorClassKey =
+    | 'container'
+    | 'icon'
+    | 'toolbar'
+    | 'title'
+    | 'panel'
+    | 'panelDetails';
+
+export interface ErrorProps {
+    classes?: Record<ErrorClassKey, string>;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Error.tsx:112]: we prefer using ClassesOverride (see my PR #4952)

File: packages/ra-ui-materialui/src/layout/Error.tsx [hunk 3]
```diff
@@ -100,12 +100,37 @@ const Error = props => {
...
+    className?: string;
+    error: Error;
+    errorInfo?: Record<'componentStack', string>;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Error.tsx:115]: where does that come from?

File: packages/ra-ui-materialui/src/layout/Error.tsx [hunk 4]
```diff
@@ -53,7 +53,7 @@ function goBack() {
     window.history.go(-1);
 }
 
-const Error = props => {
+const ErrorComponent: FC<ErrorProps> = props => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Error.tsx:56]: no, we don't name all our components XXXComponent

File: packages/ra-ui-materialui/src/layout/Error.tsx (hunk 3 above)
Comment (iamstiil) [reply packages/ra-ui-materialui/src/layout/Error.tsx:107]: It comes from https://github.com/marmelab/react-admin/blob/next/packages/ra-ui-materialui/src/layout/Layout.js#L134

File: packages/ra-ui-materialui/src/layout/Error.tsx (hunk 4 above)
Comment (iamstiil) [reply packages/ra-ui-materialui/src/layout/Error.tsx]: This is done due to the fact that the JS `Error` class is overridden in the context, but is needed as a type for the props.
